docs: Add and fix links in free-threading guide.

ngoldbaum opened the pull request:
This is mostly adding links. There are a few small text edits and fixes as well.

ngoldbaum commented:
I could probably use `PYO3_DOCS_URL` instead of `latest` in all the links to the API docs. It looks like there's only one existing link in the guide that does that, maybe I should mass-convert all the links to the API docs to do that, except if they need to point to a specific version like in the migration guide.

davidhewitt commented:
Yes please, I think it would be better to use the links to the exact PyO3 version using `PYO3_DOCS_URL`. 👍 

Other than that, this is a welcome improvement, thanks 🙏

ngoldbaum commented:
Moved all the `/latest/` links to use the `PYO3_DOCS_URL` pattern. Also rebased and added links to the docs I wrote for `OnceExt` and `OnceLockExt` in another PR.

ngoldbaum commented on guide/src/migration.md:
````diff
@@ -1064,7 +1064,7 @@ Python::with_gil(|py| {
 ```
 
-Furthermore, `Python::acquire_gil` provides ownership of a `GILGuard` which can be freely stored and passed around. This is usually not helpful as it may keep the lock held for a long time thereby blocking progress in other parts of the program. Due to the generative lifetime attached to the GIL token supplied by `Python::with_gil`, the problem is avoided as the GIL token can only be passed down  ...
+Furthermore, `Python::acquire_gil` provides ownership of a `GILGuard` which can be freely stored and passed around. This is usually not helpful as it may keep the lock held for a long time thereby blocking progress in other parts of the program. Due to the generative lifetime attached to the GIL token supplied by `Python::with_gil`, the problem is avoided as the GIL token can only be passed down  ...
````
this one is a method that will be removed in 0.23

davidhewitt commented on guide/pyclass-parameters.md:
````diff
@@ -39,7 +39,7 @@ struct MyClass {}
 ```
 
-[params-1]: https://docs.rs/pyo3/latest/pyo3/types/struct.PyAny.html
+[params-1]: {{#PYO3_DOCS_URL}}/pyo3/types/struct.PyAny.html
````
This file is actually reused inside `lib.rs` so can't have this placeholder :(

```suggestion
[params-1]: https://docs.rs/pyo3/latest/pyo3/types/struct.PyAny.html
```

davidhewitt reviewed (approved):
Super, please merge after reverting just one piece :)
